File: apps/ios/WindmillKit/Sources/WindmillGym/Ladder.swift
// Pinned against packages/api-contract/gym-ladder.json, as are web/src/products/gym/logger/ladder.js and Android's Ladder.kt.
// Bands are read off the magnitude, so the ladder behaves the same below zero.

public enum Ladder {
    public struct Band: Sendable {
        public let under: Double?
        public let small: Double
        public let large: Double
    }

    public static let bands = [
        Band(under: 20, small: 1, large: 2.5),
        Band(under: 50, small: 2.5, large: 5),
        Band(under: nil, small: 2.5, large: 10),
    ]

    // Lightening reads the band just below the magnitude, so a +1 that carried 19 → 20 is answered by −1.
    public static func steps(magnitude: Double, lightening: Bool) -> (small: Double, large: Double) {
        let band = bands[self.band(magnitude: magnitude, lightening: lightening)]
        return (band.small, band.large)
    }

    static func band(magnitude: Double, lightening: Bool) -> Int {
        bands.firstIndex { band in
            guard let under = band.under else { return true }
            return lightening ? magnitude <= under : magnitude < under
        } ?? bands.count - 1
    }

    // Half away from zero: round(−x) must equal −round(x).
    public static func round(_ weight: Double) -> Double {
        (weight * 100).rounded() / 100
    }

    public static func bump(weight: Double, direction: Int, big: Bool) -> Double {
        let step = steps(magnitude: abs(weight), lightening: Double(direction) * weight < 0)
        return round(weight + Double(direction) * (big ? step.large : step.small))
    }

    public static func labels(for weight: Double) -> [String] {
        let down = steps(magnitude: abs(weight), lightening: weight > 0)
        let up = steps(magnitude: abs(weight), lightening: weight < 0)
        return ["−\(text(down.large))", "−\(text(down.small))", "+\(text(up.small))", "+\(text(up.large))"]
    }

    // Steps print bare ("−5", not "−5.0"); the minus is U+2212, the plus ASCII.
    static func text(_ step: Double) -> String {
        if step == step.rounded() { return String(Int(step)) }
        return String(step)
    }

    // The server refuses reps < 1, and the floor clamps INTO the range rather than holding at the
    // value: a stored 0 climbs out on either key, so `bumpReps(0, -1)` is 1. `repCases` pins it.
    public static func bumpReps(_ reps: Int, direction: Int) -> Int {
        if direction < 0 { return max(1, reps - 1) }
        return reps + 1
    }
}

File: apps/android/gym/src/main/kotlin/works/windmill/gym/domain/Ladder.kt
package works.windmill.gym.domain

import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.sign

// One of three copies of this rule (apps/ios Ladder.swift, web/.../logger/ladder.js), pinned by
// packages/api-contract/gym-ladder.json — every copy reads that file as a test. Bands are read off the
// MAGNITUDE, and the step buttons do not clamp; only typed entry is bounded.

object Ladder {
    data class Band(val under: Double?, val small: Double, val large: Double)

    data class Steps(val small: Double, val large: Double)

    val bands = listOf(
        Band(under = 20.0, small = 1.0, large = 2.5),
        Band(under = 50.0, small = 2.5, large = 5.0),
        Band(under = null, small = 2.5, large = 10.0),
    )

    // A move that lightens reads the band just below the magnitude (`<` tightens to `<=`).
    fun steps(magnitude: Double, lightening: Boolean): Steps {
        // The open top band (`under: null`) answers true without comparing, so NaN lands there.
        val band = bands.firstOrNull { band ->
            val under = band.under ?: return@firstOrNull true
            if (lightening) magnitude <= under else magnitude < under
        } ?: bands.last()
        return Steps(band.small, band.large)
    }

    // Half away from zero, so round(−x) == −round(x).
    fun round(weight: Double): Double {
        val hundredths = floor(abs(weight) * 100.0 + 0.5)
        return (hundredths / 100.0) * sign(weight)
    }

    fun bump(weight: Double, direction: Int, big: Boolean): Double {
        val step = steps(abs(weight), lightening = direction * weight < 0)
        return round(weight + direction * (if (big) step.large else step.small))
    }

    fun labels(weight: Double): List<String> {
        val down = steps(abs(weight), lightening = weight > 0)
        val up = steps(abs(weight), lightening = weight < 0)
        return listOf("−${text(down.large)}", "−${text(down.small)}", "+${text(up.small)}", "+${text(up.large)}")
    }

    // The band is read the way a lift reads it (`<`): where the weight stands, not what the down key does.
    fun tier(weight: Double): String {
        val magnitude = abs(weight)
        val index = bands.indices.first { at ->
            val under = bands[at].under ?: return@first true
            magnitude < under
        }
        val band = bands[index]
        val where = when (index) {
            0 -> "under ${text(band.under ?: 0.0)} kg"
            bands.size - 1 -> "over ${text(bands[index - 1].under ?: 0.0)} kg"
            else -> "${text(bands[index - 1].under ?: 0.0)}–${text(band.under ?: 0.0)} kg"
        }
        return "$where · fine ${text(band.small)} · plate ${text(band.large)}"
    }

    // Steps print bare: "−5" and "+10", never "−5.0". The minus is U+2212, the plus ASCII.
    internal fun text(step: Double): String {
        if (step == floor(step)) return step.toInt().toString()
        return step.toString()
    }

    // Reps floor at 1: the backend refuses reps < 1, and the clamp is INTO the range.
    fun bumpReps(reps: Int, direction: Int): Int {
        if (direction < 0) return max(1, reps - 1)
        return reps + 1
    }
}
